tracer/src/Datadog.Tracer.Native/debugger_rejit_preprocessor.cpp

```diff
@@ -250,9 +276,116 @@ void DebuggerRejitPreprocessor::UpdateMethod(RejitHandlerModuleMethod* methodHan
+                                                 ComPtr<IMetaDataEmit2>& metadataEmit, const ModuleInfo& moduleInfo,
+                                                 mdTypeDef typeDef, std::vector<MethodIdentifier>& rejitRequests,
+                                                 unsigned methodDef, FunctionInfo functionInfo,
+                                                 RejitHandlerModule* moduleHandler)
+{
+    const auto [newMethodDef, newFunctionInfo] = DebuggerRejitPreprocessor::TransformKickOffToMoveNext(
+        metadataImport, metadataEmit, moduleHandler, typeDef, methodDef, functionInfo);
+    RejitPreprocessor::EnqueueNewMethod(definition, metadataImport, metadataEmit, moduleInfo, typeDef, rejitRequests,
+                                        newMethodDef, newFunctionInfo, moduleHandler);
+}
+
+std::tuple<mdMethodDef, FunctionInfo> DebuggerRejitPreprocessor::TransformKickOffToMoveNext(
+    const ComPtr<IMetaDataImport2>& metadataImport, const ComPtr<IMetaDataEmit2>& metadataEmit,
+    RejitHandlerModule* moduleHandler, mdTypeDef typeDef, mdMethodDef methodDef, const FunctionInfo& functionInfo) const
+{
+    // an alternative way is to use CustomAttributeParser
+
+    const auto iasyncStateMachineTypeRef = DebuggerRejitPreprocessor::GetIAsyncStateMachineToken(moduleHandler->GetModuleMetadata());
+    if (iasyncStateMachineTypeRef == mdTypeRefNil)
+    {
+        Logger::Error("DebuggerRejitPreprocessor::TransformKickOffToMoveNext: IAsyncStateMachine type ref token is mdTypeRefNil");
+        return {methodDef, functionInfo};
+    }
+
+    auto generatedTypeName = WStr("<") + functionInfo.name + WStr(">");
+    mdMethodDef moveNextMethod = mdMethodDefNil;
+    mdTypeDef nestedAsyncClassOrStruct = mdTypeDefNil;
+    const auto typeDefEnum = EnumTypeDefs(metadataImport);
+    auto typeDefIterator = typeDefEnum.begin();
+    for (; typeDefIterator != typeDefEnum.end(); typeDefIterator = ++typeDefIterator)
+    {
+        nestedAsyncClassOrStruct = *typeDefIterator;
+        const auto typeInfo = GetTypeInfo(metadataImport, nestedAsyncClassOrStruct);
+
+        if (!StartsWith(shared::ToString(typeInfo.name), ToString(generatedTypeName)))
+        {
+            // not a compiler generated type
+            continue;
+        }
+
+        // if it is a nested type and the parent is our type
+        mdTypeDef parentType;
+        if (metadataImport->GetNestedClassProps(nestedAsyncClassOrStruct, &parentType) != S_OK || parentType != typeDef)
+        {
+            // not a nested type or it is nested but the parent type is different from what we are looking for
+            continue;
+        }
+
+        // check if the nested type implement the IAsyncStateMachine interface
+        auto hr = DebuggerMethodRewriter::IsTypeImplementIAsyncStateMachine(metadataImport, nestedAsyncClassOrStruct);
+
+        if (hr == S_OK)
+        {
+            // only one MoveNext exist in the state machine so no need to include signature
```

Comment:
```suggestion
            // only one method called "MoveNext" exists in the state machine so no need to include the signature
```

Reply:
Fixed in 871a652d9b150d63443d87a542e34d4643145adf